nixos/modules/services/web-apps/opencloud.nix

```diff
@@ -0,0 +1,242 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  inherit (lib) types;
+  cfg = config.services.opencloud;
+
+  defaultUser = "opencloud";
+  defaultGroup = defaultUser;
+
+  settingsFormat = pkgs.formats.yaml { };
+in
+{
+  options = {
+    services.opencloud = {
+      enable = lib.mkEnableOption "OpenCloud";
+
+      package = lib.mkPackageOption pkgs "opencloud" { };
+      webPackage = lib.mkPackageOption pkgs [ "opencloud" "web" ] { };
+      idpWebPackage = lib.mkPackageOption pkgs [ "opencloud" "idp-web" ] { };
+
+      user = lib.mkOption {
+        type = types.str;
+        default = defaultUser;
+        example = "mycloud";
+        description = ''
+          The user to run OpenCloud as.
+          By default, a user named `${defaultUser}` will be created whose home
+          directory is [](#opt-services.opencloud.stateDir).
+        '';
+      };
+
+      group = lib.mkOption {
+        type = types.str;
+        default = defaultGroup;
+        example = "mycloud";
+        description = ''
+          The group to run OpenCloud under.
+          By default, a group named `${defaultGroup}` will be created.
+        '';
+      };
+
+      address = lib.mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "Web server bind address.";
+      };
+
+      port = lib.mkOption {
+        type = types.port;
+        default = 9200;
+        description = "Web server port.";
+      };
+
+      url = lib.mkOption {
+        type = types.str;
+        default = "https://localhost:9200";
+        example = "https://cloud.example.com";
+        description = "Web interface root public URL, including scheme and port (if non-default).";
+      };
+
+      stateDir = lib.mkOption {
+        default = "/var/lib/opencloud";
+        type = types.str;
+        description = "OpenCloud data directory.";
+      };
+
+      settings = lib.mkOption {
+        type = lib.types.attrsOf settingsFormat.type;
+        default = { };
+        description = ''
```

Comment:
An `example` would be nice.

Reply:
Maybe one that sets `opencloud` to whatever is in the config by default?

Reply:
That contains basically just a huge pile of secrets.

Reply:
Maybe then we should store it in /var/lib/opencloud?

Reply:
All the config has to be in one directory, unfortunately.

Reply:
But if there is no example of config that should be set via `settings` why expose that at all rather than putting all of it in /var/lib/opencloud?

Reply:
There can be an example, it's just not a very useful one unless you already understand the config system, at which point you probably don't need an example.